x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/registry/ModelRegistryMetadataTests.java

```diff
@@ -106,4 +109,145 @@ public void testAlreadyUpgraded() {
+        var newMetadata = metadata.withAddedModel(newInferenceId, newSettings);
+        assertThat(
+            newMetadata,
+            is(
+                new ModelRegistryMetadata(
+                    ImmutableOpenMap.builder(Map.of(inferenceId, settings, newInferenceId, newSettings)).build(),
+                    new HashSet<>()
+                )
+            )
+        );
+    }
+
+    public void testWithAddedModels_ReturnsSameMetadataInstance() {
+        var inferenceId = "id";
+        var settings = MinimalServiceSettingsTests.randomInstance();
+
+        var models = new HashMap<>(Map.of(inferenceId, settings));
+        var metadata = new ModelRegistryMetadata(ImmutableOpenMap.builder(models).build());
+
+        var newMetadata = metadata.withAddedModels(
+            List.of(new ModelRegistry.ModelAndSettings(inferenceId, settings), new ModelRegistry.ModelAndSettings(inferenceId, settings))
+        );
+        assertThat(newMetadata, sameInstance(metadata));
+        assertThat(newMetadata, is(new ModelRegistryMetadata(ImmutableOpenMap.builder(models).build())));
+    }
+
+    public void testWithAddedModels_ReturnsSameMetadataInstance_MultipleEntriesInMap() {
+        var inferenceId = "id";
+        var inferenceId2 = "id2";
+        var settings = MinimalServiceSettingsTests.randomInstance();
+
+        var models = new HashMap<>(Map.of(inferenceId, settings, inferenceId2, settings));
+        var metadata = new ModelRegistryMetadata(ImmutableOpenMap.builder(models).build());
+
+        var newMetadata = metadata.withAddedModels(
+            List.of(
+                new ModelRegistry.ModelAndSettings(inferenceId, settings),
+                new ModelRegistry.ModelAndSettings(inferenceId, settings),
+                new ModelRegistry.ModelAndSettings(inferenceId2, settings)
+            )
+        );
+        assertThat(newMetadata, sameInstance(metadata));
+        assertThat(newMetadata, is(new ModelRegistryMetadata(ImmutableOpenMap.builder(models).build())));
+    }
+
+    public void testWithAddedModels_ReturnsNewMetadataInstance_ForNewInferenceId() {
+        var inferenceId = "id";
+        var settings = MinimalServiceSettingsTests.randomInstance();
+
+        var models = new HashMap<>(Map.of(inferenceId, settings));
+        var metadata = new ModelRegistryMetadata(ImmutableOpenMap.builder(models).build());
+
+        var inferenceId2 = "new_id";
+        var settings2 = MinimalServiceSettingsTests.randomInstance();
+        var inferenceId3 = "new_id2";
+        var settings3 = MinimalServiceSettingsTests.randomInstance();
+        var newMetadata = metadata.withAddedModels(
+            List.of(
+                new ModelRegistry.ModelAndSettings(inferenceId2, settings2),
+                // This should be ignored since it's a duplicate
+                new ModelRegistry.ModelAndSettings(inferenceId2, settings2),
+                new ModelRegistry.ModelAndSettings(inferenceId3, settings3)
```

Comment:
Would it be worth having a test for the case where the list contains two objects with the same inference ID but different settings? We would expect the second one in the list to overwrite the first one.